cachi2/core/package_managers/rubygems/main.py

```diff
@@ -0,0 +1,98 @@
+import logging
+from typing import Optional
+
+from cachi2.core.models.input import Request
+from cachi2.core.models.output import EnvironmentVariable, ProjectFile, RequestOutput
+from cachi2.core.models.sbom import Component
+from cachi2.core.package_managers.rubygems.parser import (
+    BundlerDependency,
+    GemDependency,
+    GitDependency,
+    PathDependency,
+    parse_gemlock,
+)
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def fetch_rubygems_source(request: Request) -> RequestOutput:
+    """Resolve and process all rubygems packages."""
+    components: list[Component] = []
+    environment_variables: list[EnvironmentVariable] = []
+    project_files: list[ProjectFile] = []
+
+    for package in request.packages:
+        package_dir = request.source_dir.join_within_root(package.path)
+        _resolve_rubygems(package_dir=package_dir, output_dir=request.output_dir)
+
+    return RequestOutput.from_obj_list(
+        components=components,
+        environment_variables=environment_variables,
+        project_files=project_files,
+    )
+
+
+def _resolve_rubygems(package_dir: RootedPath, output_dir: RootedPath) -> list[Component]:
+    """Process a request for a single rubygems package."""
+    dependencies = parse_gemlock(package_dir)
+    _download_dependencies(output_dir, dependencies)
+    return _create_components(package_dir, dependencies)
+
+
+def _create_components(
+    package_dir: RootedPath,
+    dependencies: list[BundlerDependency],
+) -> list[Component]:
+    """TODO."""
+    return NotImplemented
+
+
+def _generate_purl_for(gem: BundlerDependency) -> str:
+    """TODO."""
+    return NotImplemented
+
+
+def _get_main_package_metadata(
+    package_dir: RootedPath,
+    path_dependencies: list[PathDependency],
+) -> tuple[str, Optional[str]]:
+    """TODO."""
+    return NotImplemented
+
+
+def _get_repository_name(package_dir: RootedPath) -> str:
+    """TODO."""
+    return NotImplemented
+
+
+def _download_dependencies(output_dir: RootedPath, dependencies: list[BundlerDependency]) -> None:
+    """TODO."""
+
+
+def _download_rubygems_package(output_dir: RootedPath, gem: GemDependency) -> None:
+    """TODO."""
+
+
+def _download_git_package(output_dir: RootedPath, gem: GitDependency) -> None:
+    """
+    TODO.
+
+    # directory has a specific format: {repository_name}-{revision}
+    # revision must have 12 characters
+    # https://github.com/rubygems/rubygems/blob/3da9b1dda0824d1d770780352bb1d3f287cb2df5/bundler/lib/bundler/source/git.rb#L327
+    """
+
+
+def _bundle_config_exists(package_dir: RootedPath) -> bool:
+    """TODO."""
+    return NotImplemented
+
+
+def _inject_bundle_config(path: RootedPath) -> None:
```

Comment:
During this step of the processing, I think what this function will actually do is `_generate_project_files`, since the injection will be done by another CLI command.

Reply:
I know we had a similar conversation earlier,  but what will we do if there is no `.bundle/config` in the repo?
It should be enough to just set environment variables, right ?

Reply:
(but I removed this function)

Reply:
Yeah, but then, we will need to have both behaviors: injecting project files in case there is a `.bundle/config`, and setting env vars in case there isn't. Whereas only injecting project files works either way.

As we discussed earlier, we can decide this when the implementation story comes up, but either way, this part of the code will either `generate_env_vars` or `generate_project_files`. Injection comes at a later point.

Reply:
I kept both functions for the record.